glossary.Rmd

```diff
@@ -0,0 +1,163 @@
+}"
+parse(text = code, keep.source = TRUE)
+```
+
+Source references are mostly useful for debugging and development tools. They don't play any computational role and tidy evaluation doesn't make use of references. Consequently the parsing tools in rlang return normal lists of expressions (in the rlang sense) instead of expression vectors:
+
+```{r}
+rlang::parse_exprs(code)
+```
+
+
+### TODO Symbol {#glossary-sym}
+
+
+## Programming Concepts
+
+### Constant versus symbolic {#glossary-constant-symbolic}
+
+Constants, also called "literals", always have the same value no matter the context. On the other hand symbols and calls are [symbolic](#glossary-constant-symbolic) objects: their value depends on an [environment](#glossary-env) and what kind of objects are defined there.
+
+For instance the string `"mickey"` always represents the same string no matter the environment and what objects are defined there:
+
+```{r}
+# Here's a string:
+"mickey"
+
+mickey <- "mouse"
+
+# Still the same string:
+"mickey"
+```
+
+On the other [symbols](#glossary-sym) depend on current definitions:
+
+```{r}
+# We've defined `mickey` as "mouse"
+mickey
+
+mickey <- "mickey"
+
+# Now `mickey` is "mickey"
+mickey
+```
+
+One source of problem when you're working with quoted expressions is that they might be evaluated in arbitrary places, where objects have potentially been redefined to something completely different than expected. This is a common issue with tidyverse grammars because they evaluate quoted expressions in a [data mask](#glossary-data-mask). Say you'd like to divide a column by a factor defined in the current environment:
+
+```{r}
+factor <- 100
+
+starwars %>% mutate(height / factor) %>% pull()
+```
+
+This works fine but what if the data frame contains a column called `factor`? The expression will be evaluated with the wrong definition:
+
+```{r}
+# Derive a data frame that contains a `factor` column
+starwars2 <- starwars %>% mutate(factor = 1:n())
+
+# Oh no! We're now dividing `height` by the new column!
+starwars2 %>% mutate(height / factor) %>% pull()
+```
+
+Masking is generally not a problem in scripts because you know what columns are inside your data frame. However as soon as your code is getting more general, for instance if you create a reusable function, you can no longer make assumptions about what's in the data.
+
+Fortunately with [quasiquotation](#glossary-qq) it is easy to solve masking issues by replacing symbols with constants. The unquoting operator `!!` allows you to inline constant values deep inside expressions. With [`qq_show()`](#toolbox-qq-show) we can observe the inlining:
+
+```{r}
+vector <- 1:3
+
+# Without inlining, the expression depends on the value of `vector`:
+rlang::qq_show(list(vector))
+
+# Let's inline the current value of `vector` by unquoting it:
+rlang::qq_show(list(!!vector))
+```
+
+Because constants have the same value in any environment, the data mask can never take over with parasite definitions:
```

Comment:
I love the term "parasite definitions" maybe mention it when you give the factor-as-column-name example first.